src/compiler/factory/nodeFactory.ts

```diff
@@ -4723,11 +4724,12 @@ export function createNodeFactory(flags: NodeFactoryFlags, baseFactory: BaseNode
     }
 
     // @api
-    function createImportClause(isTypeOnly: boolean, name: Identifier | undefined, namedBindings: NamedImportBindings | undefined): ImportClause {
+    function createImportClause(isTypeOnly: boolean, name: Identifier | undefined, namedBindings: NamedImportBindings | undefined, phase: ImportPhase): ImportClause {

```

Comment:
Preemptively noting that this is a breaking API change, and would require a deprecation helper in the deprecations project to tack onto our public API something that will set a default for the new parameter.

Reply:
It may also be the case that `phase` needs to go after `isTypeOnly` since AST node builder parameters and properties are intended to be in source order.

Reply:
I don't think we want to mix `type` imports with `defer` or `source`. Something like `import type source from "foo"` doesn't make sense if all `source` imports will have the same type, and `import type defer * as foo from "foo"` would essentially be the same as `import type * as foo from "foo"`.

If we consider `type`, `defer`, and `source` to be mutually exclusive, then I would suggest we replace the `isTypeOnly` parameter with something like `importModifier: SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword | SyntaxKind.SourceKeyword | boolean | undefined` and have `ImportClause` be: 

```ts
export interface ImportClause extends NamedDeclaration {
    readonly kind: SyntaxKind.ImportClause;
    readonly parent: ImportDeclaration | JSDocImportTag;
    /** @deprecated */
    readonly isTypeOnly: boolean;
    readonly importModifier: SyntaxKind.TypeKeyword | SyntaxKind.DeferKeyword | SyntaxKind.SourceKeyword | undefined;
    readonly name?: Identifier; // Default binding
    readonly namedBindings?: NamedImportBindings;
}
```

We can set `isTypeOnly` to `true` if `importModifier` is `SyntaxKind.TypeKeyword`.

Reply:
I think the issue with that scheme is that it's a little weirder to capture multiple modifiers being used in the cases of error recovery.

Reply:
I was originally going with something very similar to Ron's suggestion, and consider "type" just as another phase. I ended up not doing it because of the AST breaking change, but if just keeping the old property for backwards compat is ok then I'd go for it.

> I think the issue with that scheme is that it's a little weirder to capture multiple modifiers being used in the cases of error recovery.

Error recovery is going to be tricky anyway, because each modifier is a valid identifier so things like `import type defer` are a potentially valid start of an import declaration. I wonder how likely it is for people to accidentally write two modifiers in an import.